### ISSUE TITLE
Implementing packbits

### ISSUE THREAD

**Issue by aayn** [pytorch/pytorch#42223]
Requested in #32867 and #38349 (NumPy rollup).

cc @mruberry

**Comment by aayn** [comment]
@mruberry What would be right place to put this?

**Comment by mruberry** [comment]
@aayn https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/TensorFactories.cpp

**Comment by aayn** [comment]
From #32867:
> @aayn numpy's packbkits is quite restricted and supports packing only BoolTensors. I propose here to expand its functionalities to custom masks and arbitrary data types.
> 
> Need for fully-fledged BitTensors also needs some discussion. Same for relation with existing 2-bit quantization and quantization in general, since BitTensors and low-precision quantization could be used for binary neural net works.

@mruberry @vadimkantorov Does it make more sense to emulate NumPy first and then expand its scope, or to write the more general version straight up?

**Comment by mruberry** [comment]
> From #32867:
> 
> > @aayn numpy's packbkits is quite restricted and supports packing only BoolTensors. I propose here to expand its functionalities to custom masks and arbitrary data types.
> > Need for fully-fledged BitTensors also needs some discussion. Same for relation with existing 2-bit quantization and quantization in general, since BitTensors and low-precision quantization could be used for binary neural net works.
> 
> @mruberry @vadimkantorov Does it make more sense to emulate NumPy first and then expand its scope, or to write the more general version straight up?

Focusing on implementing the function like NumPy does sounds great. Further potential enhancements should be separable.

**Comment by vadimkantorov** [comment]
Also, NumPy packbits API for undoing padding is a little unintuitive. Even PyTorch fft `signal_sizes` is more intuitive.

I really wish PyTorch gets a more usable, modernized version of packbits / quantization as well. Of course it can be a separate improvement.

**Comment by aayn** [comment]
I'd be interested to work to the improvement after this too. For some reason, I'm having issues building PyTorch from source right now.

**Comment by aayn** [comment]
@mruberry What's the best/recommended way to do reduction along a particular `dim` within PyTorch? This is so that I can packbits along that dimension.

**Comment by mruberry** [comment]
TensorIterator has code to make generic reductions, but packbits is not really a reduction, like sum, since it does not involve all the values in an axis but rather 8 element segments of that access.

For example, 

```
a = np.array([[1, 0], [0, 1], [0, 1], [0, 1], [0, 1], [0, 1], [0, 1], [0, 1], [0, 1], [0, 1]])
: array([[1, 0],
       [0, 1],
       [0, 1],
       [0, 1],
       [0, 1],
       [0, 1],
       [0, 1],
       [0, 1],
       [0, 1],
       [0, 1]])
```

Here the array is 2D 10x2 (10 rows and 2 columns). If we packbits along axis 0 we get:

```
np.packbits(a, axis=0)
: array([[128, 127],
       [  0, 192]], dtype=uint8)
```

Which is a 2x2 matrix. If we had taken the sum of a, however, we'd get:

```
np.sum(a, axis=0)
: array([1, 9])
```

A 1x2 matrix, because the first axis has been reduced.

Packbits is more like a windowing function, like those found in TensorFactories:

https://github.com/pytorch/pytorch/blob/2f840b1662b487d5551d7230f8eb4d57645cfff5/aten/src/ATen/native/TensorFactories.cpp#L817

An optimal implementation of packbits is really interesting, but also tricky, I think, because you want to enumerate these 8 element blocks of the input tensor in a particular order to compute the output values. An optimal implementation would also account for caching. For example, for a large matrix if you wanted to pack the columns an optimal kernel would read a 2D section of the matrix, transpose it in (shared) memory, then compute the bit vector values. 

That would be a challenging function to implement that, I think, would require custom CPU and CUDA kernels. In fact, we might want to make the ability to compute windowing functions like packbits generic, which ups the challenge a little. That is, we might want to build the architecture to make it easy to write functions like packbits in a perf-optimal (or near perf-optimal) way, where a developer just has to specify the axis, the window size, and the computation over that window. Analogous to how TensorIterator makes reductions easy to write today.

Another way to approach this function, however, is to see if it's reasonable to write using existing PyTorch operations, possibly with a simpler helper kernel. For example, you could transpose and copy your input so you're always packing bits along contiguous rows, run a windowing kernel on that tensor, then transpose the output of your kernel back to the correct original form. This creates an extra copy of the input and possibly an extra copy of the output, but that might be OK for packbits. You could even make this approach more generic and offer that as an initial "generic window function" architecture, depending on how it goes.

Alternatively, if you wanted to work on a proper reduction instead of a windowing function, I think there are a few of those to implement, too.

**Comment by vadimkantorov** [comment]
Fast "shift and reduce" draft implementations are available in https://github.com/pytorch/pytorch/issues/32867#issuecomment-662619393 However, they only work for padded tensors. Given that this is about perf optimizations in the end, it would be nice to avoid copies...

**Comment by aayn** [comment]
I still want to go ahead with implementing packbits. Given that it's challenging, I think I'll be able to learn a lot from this. 

Point taken, will avoid copies to the extent possible.

**Comment by mruberry** [comment]
Avoiding copies is always good but I wouldn't go crazy over it. Other functions also sometimes have to copy their inputs and outputs, especially if the data comes in discontiguous.

**Comment by vadimkantorov** [comment]
In principle - sure. I mean for the case when input/output are contiguous

**Comment by vadimkantorov** [comment]
For CPU impl, expand/compress SSE/AVX instructions might be related

**Comment by aayn** [comment]
@mruberry I wrote a naive implementation of packbits (for my own understanding) and now I'm looking to write a more optimized version. Is there something like NumPy's parallel [iterating over all but one axis](https://github.com/numpy/numpy/blob/558080ef33de051c52bd1214c372380680cb6800/doc/source/user/c-info.beyond-basics.rst#iterating-over-all-but-one-axis) in PyTorch?

**Comment by mruberry** [comment]
> @mruberry I wrote a naive implementation of packbits (for my own understanding) and now I'm looking to write a more optimized version. Is there something like NumPy's parallel [iterating over all but one axis](https://github.com/numpy/numpy/blob/558080ef33de051c52bd1214c372380680cb6800/doc/source/user/c-info.beyond-basics.rst#iterating-over-all-but-one-axis) in PyTorch?

Take a look at parallel.h and in particular:

https://github.com/pytorch/pytorch/blob/ccea3726da3aca338eb7753e84ba370a13148855/aten/src/ATen/Parallel.h#L113

Would something like that work?

**Comment by aayn** [comment]
Thanks! I think `parallel_for` might work. But I think these parallel utilities work only if the input is padded, right?. That is, only if the dimension along which we are packing is a multiple of 8. Do you think it is worth the tradeoff to pad (with zeros) the input first and then do the packing? This would probably take up some extra memory, but could potentially be faster and easier to implement.
@mruberry

**Comment by mruberry** [comment]
In general these functions do work on unpadded inputs (`torch.sum` uses them, for example) but it's likely you'd have to create modified versions of them to do what you'd like (that is, do something along the lines of creating helpers for windowing functions).

To recap our earlier discussion, I think there are still four ways to go:

- develop generic helpers for "windowing functions" like packbits by doing something similar to these parallel functions and TensorIterator's use of them (reviewing how pooling functions like maxpool1d use them might be interesting, too)
- write custom kernels
- see if there's a performant implementation mixing existing Torch ops and simpler custom kernels
- take a look at other functions (or tasks) better supported by PyTorch's existing architecture

As mentioned earlier, this is a tricky direction to go in at this time.

**Comment by aayn** [comment]
I really wanted to do something tricky/challenging, but my schoolwork has been too demanding for the last few weeks. I think I'll take a break for now, and come back when things are more under control. Thank you so much, and sorry for taking up so much of your time for naught! Feel free to close this issue/reassign it so that anyone who's interested.
@mruberry

**Comment by mruberry** [comment]
> I really wanted to do something tricky/challenging, but my schoolwork has been too demanding for the last few weeks. I think I'll take a break for now, and come back when things are more under control. Thank you so much, and sorry for taking up so much of your time for naught! Feel free to close this issue/reassign it to anyone who's interested.
> @mruberry

Of course. But please don't think you wasted anyone's time or it's been for nothing. If we only picked projects that were easy we would never learn or develop, and I've appreciated our conversations about packbits. Eventually we will implement this function, and this issue will be helpful in that process.

**Comment by aayn** [comment]
Wow, thanks again!

**Comment by bruce2233** [comment]
The issue still opens.
